### PR TITLE
[FEATURE] Migrate user rights

### PR DESCRIPTION
**Pull request by hschink**
Implements [ticket 1987](https://www.openproject.org/issues/1987)

:exclamation: This PR contains #441

### CODE REVIEW

File: db/migrate/20130917101922_migrate_query_tracker_references_to_type.rb [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+  def down
+    update_tracker_references_with_keys(KEY.invert)
+  end
+
+  private
+
+  def update_tracker_reference(keys, columns)
+    Proc.new do |row|
+      COLUMNS.each do |column|
```
Comment (ulferts) [inline db/migrate/20130917101922_migrate_query_tracker_references_to_type.rb]: I believe we misunderstood one another here. You should either write:

> def update_tracker_reference(keys, columns)
>    Proc.new do |row|
>      columns.each do |column| 

if columns is to be provided as part of the method call, or you should write:

> def update_tracker_reference(keys)
>    Proc.new do |row|
>      COLUMNS.each do |column| 

if columns will always be COLUMNS.

Currently "columns" needs to be provided without it being used in the method.

File: db/migrate/20130917101922_migrate_query_tracker_references_to_type.rb [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+          row[column] = YAML.dump value
+        end
+      end
+
+      row
+    end
+  end
+
+  def update_tracker_references_with_keys(keys)
+    filter = COLUMNS.join(" LIKE '%tracker%' OR ") + " LIKE '%tracker%'"
```
Comment (ulferts) [inline db/migrate/20130917101922_migrate_query_tracker_references_to_type.rb]: Just a question of style I believe, but I would prefer:

> filter = COLUMNS.map{ |c| "#{c} LIKE '%tracker%'" }.join("  OR  ")

File: db/migrate/20130918111753_migrate_user_rights.rb [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    view_issue_watchers: :view_work_package_watchers,
+    add_issue_watchers: :add_work_package_watchers,
+    delete_issue_watchers: :delete_work_package_watchers,
+
+    view_planning_elements: :view_work_packages,
+    edit_planning_elements: :edit_work_packages,
+    delete_planning_elements: :delete_work_packages,
+
+    issue_form_issue: nil,
+    manage_issue_forms: nil,
```
Comment (ulferts) [inline db/migrate/20130918111753_migrate_user_rights.rb]: those two stem from a plugin, they should not be touched.

Comment (apfelfabrik) [inline db/migrate/20130918111753_migrate_user_rights.rb]: If they're from plugins, why mention them at all?

Comment (ulferts) [inline db/migrate/20130918111753_migrate_user_rights.rb]: That is what I meant. Thanks for clarifying it.

File: db/migrate/20130918111753_migrate_user_rights.rb [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    edit_own_issue_notes: nil,
```
Comment (ulferts) [inline db/migrate/20130918111753_migrate_user_rights.rb]: You replaced this with :edit_own_work_package_notes in the code, so this should also be the replacement here.

File: db/migrate/migration_utils/utils.rb [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+      select_rows_for_update(table, column_list, conditions).each do |row|
+        updated_rows << updater.call(row)
+      end
+
+      update_rows_in_database(table, column_list, updated_rows)
+    end
+
+    private
+
+    def select_rows_for_update(table, column_list, conditions)
```
Comment (ulferts) [inline db/migrate/migration_utils/utils.rb]: The method name should not state why you call this but rather what you expect the method to do. The "for_update" part is wrong here.

File: db/migrate/migration_utils/utils.rb [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    end
+
+    def update_rows_in_database(table, column_list, updated_rows)
+      columns = (column_list.nil?) ? '' : column_list.join(', ')
+
+      updated_rows.each do |row|
+        values = column_list.map {|c| "#{c}=#{quote(row[c])}"}
+                            .join(', ')
+
+        insert <<-SQL
```
Comment (ulferts) [inline db/migrate/migration_utils/utils.rb]: this is not an insert, its an update. The benefit of that is that the number of updated rows is returned. Something the method could return.

File: db/migrate/20130918111753_migrate_user_rights.rb [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+    view_issue_watchers: :view_work_package_watchers,
+    add_issue_watchers: :add_work_package_watchers,
+    delete_issue_watchers: :delete_work_package_watchers,
+
+    view_planning_elements: :view_work_packages,
+    edit_planning_elements: :edit_work_packages,
+    delete_planning_elements: :delete_work_packages,
+
+    edit_own_issue_notes: nil,
```
Comment (ulferts) [inline db/migrate/20130918111753_migrate_user_rights.rb]: > You replaced this with :edit_own_work_package_notes in the code, so this should also be the replacement here.

still applies

**Comment by ulferts** [comment]
There are still remains of issue permissions, e.g.:
- view_issues
- edit_issue
- add_issue

Could you please double check for all removed permissions, and remove/replace what comes up.

**Pull request merged**